sdk/engine/source/assets/TextureAssetManager.cpp

```diff
@@ -13,127 +9,153 @@
-    }
+    m_onUpdate.Emit
+    (
+        TextureBufferData(UpdateAction::Load, std::vector<std::string>{path}, std::span<const TextureData>{&m_textureData.back(), 1})
+    );
+    return true;
+}
 
-    bool TextureAssetManager::Load(std::span<const std::string> paths, bool isExternal)
-    {
-        const auto newTextureCount = static_cast<uint32_t>(paths.size());
-        auto nextTextureIndex = static_cast<uint32_t>(m_textures.size());
-        if (newTextureCount + nextTextureIndex >= m_maxTextureCount)
-            throw NcError("Cannot exceed max texture count.");
+bool TextureAssetManager::Load(std::span<const std::string> paths, bool isExternal)
+{
+    const auto newItemsIndex = m_accessors.size();
+    const auto newTextureCount = static_cast<uint32_t>(paths.size());
+    auto nextTextureIndex = static_cast<uint32_t>(newItemsIndex);
+    if (newTextureCount + nextTextureIndex >= m_maxTextureCount)
+        throw NcError("Cannot exceed max texture count.");
 
-        for (const auto& path : paths)
-        {
-            if(IsLoaded(path))
-                continue;
+    auto idsToLoad = std::vector<std::string>{};
+    idsToLoad.reserve(paths.size());
 
-            const auto fullPath = isExternal ? path : m_assetDirectory + path;
-            m_textures.push_back(ReadTexture(fullPath, m_graphics, m_sampler.get()));
-            m_accessors.emplace(path, nextTextureIndex++);
+    for (const auto& path : paths)
+    {
+        if(IsLoaded(path))
+        {
+            continue;
         }
-
-        graphics::ShaderResourceService<graphics::Texture>::Get()->Update(m_textures);
-        return true;
+        
+        m_accessors.emplace_back(nextTextureIndex++);
+        const auto fullPath = isExternal ? path : m_assetDirectory + path;
+        m_textureData.push_back(ReadTexture(fullPath, path));
+        idsToLoad.push_back(path);
     }
 
-    bool TextureAssetManager::Unload(const std::string& path)
+    if (!idsToLoad.empty())
     {
-        auto removed = static_cast<bool>(m_accessors.erase(path));
-        if(!removed)
-            return false;
-        
-        auto pos = std::ranges::find_if(m_textures, [&path](const auto& texture)
-        {
-            return texture.uid == path;
-        });
-
-        assert(pos != m_textures.end());
-        auto index = std::distance(m_textures.begin(), pos);
-        m_textures.erase(pos);
+        m_onUpdate.Emit
+        (
+            TextureBufferData(UpdateAction::Load, std::move(idsToLoad), std::span<const TextureData>{m_textureData.begin() + newItemsIndex, m_textureData.end()})
+        );
+    }
+    return true;
+}
 
-        for(auto& [unused, textureView] : m_accessors)
-        {
-            if(textureView.index > index)
-                --textureView.index;
-        }
+bool TextureAssetManager::Unload(const std::string& path)
+{
+    auto pos = std::ranges::find_if(m_textureData, [&path](const auto& data)
+    {
+        return data.id == path;
+    });
 
```

Comment:
Looking at this, I'm not sure m_accessors is really even needed anymore. I think its always an ordered list of indices. We can figure out everything we need from m_textureData.

Reply:
Was the idea for m_accessors to be a sparse set type thing? Maybe that's what's missing. Some discussion may be required here.